### PR TITLE
CI: Add ccache

### PR DESCRIPTION
**Pull request by jonashaag**
Add Sccache https://github.com/mozilla/sccache to the GHA CI. Saves around 5 minutes per run.

Unfortunately it requires some hacks, namely changing a temporary Windows directory and patching distutils.

- [ ] closes #xxxx (Replace xxxx with the Github issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: ci/run_tests.sh [hunk 1]
```diff
@@ -1,4 +1,6 @@
 #!/bin/bash -e
+python -c 'import pandas; pandas.show_versions()'
+exit 0
```
Comment (jonashaag) [inline ci/run_tests.sh:3]: Obviously this is only temporary to increase PR iteration speed.

File: .github/actions/build-pandas/action.yml [hunk 2]
```diff
@@ -5,14 +5,23 @@ runs:
   steps:
 
     - name: Environment Detail
+      id: env-detail
       run: |
+        echo ::group::{Environment Detail}
         conda info
         conda list
+        echo ::set-output name=python-version::$(conda list -f python --json | jq -r '.[0].version')
```
Comment (mroeschke) [inline .github/actions/build-pandas/action.yml:13]: The conda environment is activate by the time this runs correct?

Comment (jonashaag) [reply .github/actions/build-pandas/action.yml:13]: Yep!

File: .github/actions/setup-ccache/action.yml [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+      # may become read-only at some point. Work around by having a private tempdir.
+      id: mktemp
+      run: echo "::set-output name=tmpdir::$(cygpath -w $(mktemp -d))"
+      shell: bash
+      if: ${{ runner.os == 'Windows' }}
+
+    - name: Setup sccache
+      uses: hendrikmuhs/ccache-action@v1.2
+      with:
+        variant: sccache
```
Comment (mroeschke) [inline .github/actions/setup-ccache/action.yml:25]: From browsing sccache, it seems like the only nice benefit is flexible cloud provider storage options which is not relevant here?

**Review by mroeschke — COMMENTED** [review]
Might need some documentation on how to manually invalidate the cache. We can potentially get into a bad state like e.g. https://github.com/pandas-dev/pandas/issues/47305

File: .github/actions/setup-ccache/action.yml (hunk 3 above)
Comment (jonashaag) [reply .github/actions/setup-ccache/action.yml:25]: Ccache has only recently gained Windows (MSVC) support . I’m not sure how stable it is but we can try. Sccache Windows support is older and my hypothesis is that this also means it’s more stable.

**Comment by jonashaag** [comment]
> Might need some documentation on how to manually invalidate the cache. We can potentially get into a bad state like e.g. #47305

Done!

**Comment by mroeschke** [comment]
While this looks good, the `setup.py` changes may conflict with https://github.com/pandas-dev/pandas/pull/47380 fairly soon. Mind pivoting to other CI improvements (like environment caching) if you have the time until a decision has been made on the linked PR?

**Comment by jbrockmendel** [comment]
Could some of the hacks be avoided by only caching on non-windows?

**Comment by jbrockmendel** [comment]
@jonashaag status here?  i was looking forward to this one

**Comment by jonashaag** [comment]
I put this on hold to wait for the Meson changes. Are they done already?

**Comment by datapythonista** [comment]
> I put this on hold to wait for the Meson changes. Are they done already?

Seems like #49115 is almost ready, maybe a good time to update this?

**Comment by mroeschke** [comment]
Do you think the sccache-action could be easily be used for this instead? https://github.com/Mozilla-Actions/sccache-action

**Comment by jonashaag** [comment]
This is already using another action. What would be the advantage of switching actions? TBH I think the actions are both very similar

**Comment by mroeschke** [comment]
Ah I didn't know if it would make this workflow any simpler, in addition to be more of an official looking ccache action by mozilla

**Comment by mroeschke** [comment]
Closing in favor of https://github.com/pandas-dev/pandas/pull/53211

**Pull request closed without merging**